### ISSUE TITLE
SSL Handshake error doing mutual authentication

### ISSUE THREAD

**Issue by mschiessl** [greenbone/ospd-openvas#175]
Hello 

i am having issues connecting my gvmd setup to a remote scanner using the certificates created with gvmd-manage-certs. Adding the scanner via GSAD and hitting the "Verify" button, i get the following error on the remote box:

```
2019-12-17 15:29:59,390 OSPD - openvas: DEBUG: (ospd.server) New connection from ('xxx.xxx.xxx.xxx', 51080)
----------------------------------------
Exception happened during processing of request from ('xxx.xxx.xxx.xxx', 51080)
Traceback (most recent call last):
  File "/usr/lib/python3.7/socketserver.py", line 650, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/lib/python3.7/socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python3.7/socketserver.py", line 720, in __init__
    self.handle()
  File "/opt/openvas/lib/python3.7/site-packages/ospd-2.0.0-py3.7.egg/ospd/server.py", line 126, in handle
    self.server.handle_request(self.request, self.client_address)
  File "/opt/openvas/lib/python3.7/site-packages/ospd-2.0.0-py3.7.egg/ospd/server.py", line 166, in handle_request
    self.server.handle_request(request, client_address)
  File "/opt/openvas/lib/python3.7/site-packages/ospd-2.0.0-py3.7.egg/ospd/server.py", line 294, in handle_request
    req_socket = self.tls_context.wrap_socket(request, server_side=True)
  File "/usr/lib/python3.7/ssl.py", line 423, in wrap_socket
    session=session
  File "/usr/lib/python3.7/ssl.py", line 870, in _create
    self.do_handshake()
  File "/usr/lib/python3.7/ssl.py", line 1139, in do_handshake
    self._sslobj.do_handshake()
OSError: [Errno 0] Error
----------------------------------------
```

on the **remote** i copied the following certifcates:
- cacert.pem  
- servercert.pem  
- serverkey.pem  

for the **local** scanner credentials, i've used the following certs:
- cacert.pem
- clientcert.pem
- clientkey.pem

Versions on the **local side** :
```
gsa.vers
v9.0.0
gvm-libs.vers
v11.0.0
gvm-tools.vers
v2.0.0
gvmd.vers
v9.0.0
openvas-smb.vers
v1.0.5
openvas.vers
v7.0.0
ospd-openvas.vers
v1.0.0
ospd.vers
v2.0.0
```

Versions on the **remote side**:
```
./ospd-openvas --version
OSP Server for openvas: 1.0.0
OSP: 1.2
OSPd: 2.0.0

Copyright (C) 2014, 2015, 2018, 2019 Greenbone Networks GmbH
License GPLv2+: GNU GPL version 2 or later
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

All binaries / scripts are grabbed from github and compiled on the latest "KALI LINUX" version (rolling). 

Appreciate any input / suggestions

**Comment by mschiessl** [comment]
ok ... digging deeper into...
From The local side (running gvmd) 

```
root@kali# openssl s_client -connect xxx.xxx.xxx.xxx:port
 -CAfile /opt/openvas/var/lib/gvm/CA/cacert.pem -cert /opt/openvas/var/lib/gvm/CA/clientcert.pem
-key /opt/openvas/var/lib/gvm/private/CA/clientkey.pem  -tls1_2 -state -quiet
SSL_connect:before SSL initialization
SSL_connect:SSLv3/TLS write client hello
SSL_connect:SSLv3/TLS write client hello
Can't use SSL_get_servername
SSL_connect:SSLv3/TLS read server hello
depth=1 OU = Certificate Authority for gvmd-side, O = GVM Users, L = Osnabrueck, C = DE
verify return:1
depth=0 CN = k4l1-m, O = GVM Users, L = Osnabrueck, C = DE
verify return:1
SSL_connect:SSLv3/TLS read server certificate
SSL_connect:SSLv3/TLS read server key exchange
SSL_connect:SSLv3/TLS read server certificate request
SSL_connect:SSLv3/TLS read server done
SSL_connect:SSLv3/TLS write client certificate
SSL_connect:SSLv3/TLS write client key exchange
SSL_connect:SSLv3/TLS write certificate verify
SSL_connect:SSLv3/TLS write change cipher spec
SSL_connect:SSLv3/TLS write finished
SSL_connect:SSLv3/TLS write finished
SSL_connect:SSLv3/TLS read server session ticket
SSL_connect:SSLv3/TLS read change cipher spec
SSL_connect:SSLv3/TLS read finished
read:errno=0
SSL3 alert write:warning:close notify
```

so this seems to be totally fine connecting to the box with mutual authentication.
Assuming the issue is on GVMD or GSAD side adding a scanner

**Comment by mschiessl** [comment]
adding  the Scanner on **local side** via CLI:

```
/opt/openvas/sbin/gvmd --create-scanner="OSP TEST Scanner" --scanner-host=xxx.xxx.xxx.xxx --scanner-port=port \
>       --scanner-type="OSP" --scanner-ca-pub=/opt/openvas/var/lib/gvm/CA/cacert.pem \
>       --scanner-key-pub=/opt/openvas/var/lib/gvm/CA/clientcert.pem \
>       --scanner-key-priv=/opt/openvas/var/lib/gvm/private/CA/clientkey.pem
```
adds the scanner  properly.

Firing a verify against the scanner:
```
./gvmd --verify-scanner=xxxxxxxxx-xxxxxxx-xxxx-xxxxxx
Failed to verify scanner.
```

On the remote side the absolute identical error message is generated as in the error message within the top post.

this brings me to the point that its more than likely gvmd having the issue - gonna try to do some more debugging

**Comment by mschiessl** [comment]
ok played around a little further but no success:
- exchanged key and pem (input error test)
- exchanged client with server certs (same cert)
- added CA to the system CA list 
- used GSAD and CLI 
no success :( 

Debug log in GVMD:
```
util gpgme:MESSAGE:2019-12-17 19h49.04 utc:16871: Setting GnuPG dir to '/opt/openvas/var/lib/gvm/gvmd/gnupg'
util gpgme:MESSAGE:2019-12-17 19h49.04 utc:16871: Using OpenPGP engine version '2.2.17'
lib  serv:  DEBUG:2019-12-17 19h49.04 utc:16871:    Connected to server 'xxx.xxx.xxx.xxx' port port.
lib  serv:  DEBUG:2019-12-17 19h49.04 utc:16871: Failed to shake hands with server 'xxx.xxx.xxx.xxx' port port: Key usage violation in certificate has been detected.
md manage:WARNING:2019-12-17 19h49.04 utc:16871: Could not connect to Scanner at xxx.xxx.xxx.xxx:port```

So the magic error message is: **"Key usage violation in certificate has been detected."**

Please feel free to move the ticket towards GVMD or  GVM-LIBS whatever fits better.

**Comment by djordjezecevic** [comment]
Exact same problem here, just different versions. 
gvm@openvas-razvoj:~$ /opt/gvm/bin/ospd-scanner/bin/ospd-openvas --version
OSP Server for openvas: 20.4a1
OSP: 1.2
OSPd: 20.4a1

Regards,
Djordje

**Comment by mschiessl** [comment]
@djordjezecevic  - did you have any progress in getting the remote to work ? 
I was waiting for some new releases but from creation of that issue, nothing was released.
Thinking about going with the "master" branches to see if they work.

Best 
Mike

**Comment by bjoernricks** [comment]
Please don't use the master branch. It is a development branch that may break everytime! Please use the latest release branch instead.

**Comment by djordjezecevic** [comment]
@mschiessl Unfortunately no, college of mine also tried, no luck.

**Comment by mschiessl** [comment]
ok - updated the binaries from the latest release branches from yesterday night.
Still receiving the same error as stated in the initial post.
```
 File "/usr/lib/python3.7/ssl.py", line 1139, in do_handshake
    self._sslobj.do_handshake()
```

any other ideas upon this topic - would really be great to get remote ospd's working

**Comment by BasvanH** [comment]
Whats the subject / DN of your **remote** certificates? Is it the IP address you enter while creating the OSP scanner with `gvmd`? If it's the hostname of your **local**, then you should add the OSP scanner with the hostname to `gvmd`. Secondly, you need to copy the remote certificates to your local server, and add the OSP scanner with those certificates and key. I just got this working and remote is properly validated.

**Comment by mcjon3z** [comment]
@BasvanH - could you elaborate a bit more on how you got this working? I'm getting similar errors relating to SSL handshake as noted above:

1. Created slave scanner certs using gvm-manage-certs and passing the CN via environment variables to match the slave hostname which I'm connecting to.
2. Copied slave scanner certs to master server and used those certs for the following setups
3. Created OSP type scanner in GSA using the cacert, servercert, and serverpub certificates - no dice
4. Created OSP type scanner in GSA using the cacert, clientcert, and clientpub certificates - no dice
5. Used gvmd command line to create an OpenVAS type scanner using the cacert, clientcert, and clientpub certificates - no dice
6. Used gvmd command line to create an OpenVAS type scanner using the cacert, servercert, and serverpub certificates - no dice

**Comment by wisukind** [comment]
I confirm I have the same problem. Tested with various versions of gvmd / gvm-libs on various distributions. However the problem do not exist on Linux Mint 18.3; but on this version you can only install latest stable version. Latest release branches won't compile.

**Comment by wisukind** [comment]
Also @BasvanH , on my side the scanners were created with IP addresses, not hostnames. And again, it was working without problems on earlier Linux distros. What linux distribution are you using to get it work ?

**Comment by BasvanH** [comment]
I'm on Ubuntu 18.04 TLS, used the steps as described in step 4. Via GSA add OSP scanner with fqdn of remote, port 9390 and cacert. Added credential using clientcert & clientkey (no passphase) to the scanner.

On remote:
`/opt/gvm/bin/ospd-scanner/bin/python /opt/gvm/bin/ospd-scanner/bin/ospd-openvas --pid-file /opt/gvm/var/run/ospd-openvas.pid --log-file /opt/gvm/var/log/gvm/ospd-scanner.log --lock-file-dir /opt/gvm/var/run --bind-address 0.0.0.0 --port 9390 --key-file /opt/gvm/var/lib/gvm/private/CA/clientkey.pem --cert-file /opt/gvm/var/lib/gvm/CA/clientcert.pem --ca-file /opt/gvm/var/lib/gvm/CA/cacert.pem`

Install sources on both sides:
```
git clone -b gvm-libs-11.0 https://github.com/greenbone/gvm-libs.git
git clone -b gvmd-9.0 https://github.com/greenbone/gvmd.git
git clone -b gsa-9.0 https://github.com/greenbone/gsa.git
git clone https://github.com/greenbone/openvas-smb.git
git clone -b openvas-7.0 https://github.com/greenbone/openvas.git
git clone -b ospd-2.0 https://github.com/greenbone/ospd.git
git clone -b ospd-openvas-1.0 https://github.com/greenbone/ospd-openvas.git
```

**Comment by wisukind** [comment]
Thanks. I will try that now.

**Comment by wisukind** [comment]
So I confirm following the above procedure works. Eg; you can verify the scanners successfully. However it apparentely create a new issues; setting the scanner to OSP type generate an error whenever you want to create or modify a task using the created scanner. (screenshot attached)

![Screenshot at 2020-04-24 12-01-48](https://user-images.githubusercontent.com/48824957/80200773-6c77f680-8623-11ea-88ce-e0e5f6389552.png)

**Comment by wisukind** [comment]
So you can't create a new task with the newly created scanner.  But tasks created before and assigned with this scanner uuid works and can be started. So there is definitely something wrong.

**Comment by BasvanH** [comment]
In GSA you need to create a new **Scan Config** for this scanner. **Configuration** / **Scan Configs** and select your created OSP scanner.

**Comment by wisukind** [comment]
I did that, but I still have this error.  Also the newly created scan config looks empty. Is it normal ?
![Screenshot at 2020-04-24 12-15-22](https://user-images.githubusercontent.com/48824957/80202047-523f1800-8625-11ea-8b1f-2142463c3087.png)

**Comment by wisukind** [comment]
@BasvanH 
Somehow it works for newly created tasks. However, it doesn't work for existing alterable tasks for which you modify the scanner. I have a lot of tasks which were previously using a remote scanner, of type "OpenVAS"/ 
Is it possible to modify the existing OpenVAS remote scanner with gvmd --modify-scanner so that I don't have to recreate all my tasks (and loosing the previous data) ?

Also, if it's not too much asking, could you explain a little bit the logic behind the need to recreate a scan config associated with the scanner ? 

Thanks a lot for your help

**Comment by wisukind** [comment]
@BasvanH

Also one question; from https://github.com/greenbone/gvmd/pull/869/commits/ce6df87dbead84d90e0310b72e5916a5ee74795b it seems now gvmd can RESUME OSP type scans. I'm using the latest stable release, and I stil get the error:

md manage:WARNING:2020-04-24 13h33.59 CEST:1169: run_osp_scan_get_report: Scanner type does not support resuming scans
md manage:WARNING:2020-04-24 13h33.59 CEST:1169: Couldn't fork OSP scan handler

Whenever I'm trying to resume an OSP scan task ? Hope I'm not raising this topic at the wrong place. Thanks !

**Comment by BasvanH** [comment]
Yo are sort of stealing this topic and this is a place where we should discus code issues. Please create a topic on the Greenbone community forum.

**Comment by wisukind** [comment]
Will do. Sorry for that and thanks for your help !

**Comment by mcjon3z** [comment]
> I'm on Ubuntu 18.04 TLS, used the steps as described in step 4. Via GSA add OSP scanner with fqdn of remote, port 9390 and cacert. Added credential using clientcert & clientkey (no passphase) to the scanner.
> 
> On remote:
> `/opt/gvm/bin/ospd-scanner/bin/python /opt/gvm/bin/ospd-scanner/bin/ospd-openvas --pid-file /opt/gvm/var/run/ospd-openvas.pid --log-file /opt/gvm/var/log/gvm/ospd-scanner.log --lock-file-dir /opt/gvm/var/run --bind-address 0.0.0.0 --port 9390 --key-file /opt/gvm/var/lib/gvm/private/CA/clientkey.pem --cert-file /opt/gvm/var/lib/gvm/CA/clientcert.pem --ca-file /opt/gvm/var/lib/gvm/CA/cacert.pem`
> 
> Install sources on both sides:
> 
> ```
> git clone -b gvm-libs-11.0 https://github.com/greenbone/gvm-libs.git
> git clone -b gvmd-9.0 https://github.com/greenbone/gvmd.git
> git clone -b gsa-9.0 https://github.com/greenbone/gsa.git
> git clone https://github.com/greenbone/openvas-smb.git
> git clone -b openvas-7.0 https://github.com/greenbone/openvas.git
> git clone -b ospd-2.0 https://github.com/greenbone/ospd.git
> git clone -b ospd-openvas-1.0 https://github.com/greenbone/ospd-openvas.git
> ```

@BasvanH - thank you for the reply. I believe I see the issue. My slave was using the generated server certs instead of client certs based on the default config file. In looking at the source code for gvm-manage-certs, there are different generation parameters for certificate use between the server and client cert types. This makes sense based on some of the key usage errors that were identified by OP in this thread and the related one above. I will test later this morning and post an update.

**Comment by mcjon3z** [comment]
As an update, I can confirm that using the setup as described above is now allowing verification of the remote scanner without SSL errors using the clientcert certificate.

TLDR - when generating the certificates for the remote scanner, you must specify the CN so that it matches whatever IP/hostname/FQDN you will be connecting to the scanner from AND the certificate type on the remote scanner must be a client certificate, not a server certificate. If either of these are not configured correctly, you will receive SSL handshake errors as described above.

**Comment by BasvanH** [comment]
Nice to hear you got it working.

**Comment by jjnicola** [comment]
Fixed with greenbone/ospd/#284 and greenbone/gvmd#1139

**Comment by sebasdelho** [comment]
Hi Matthew C, i can help me with Openvas I have an error when verifying the scanner "Failed to verify scanner" I am using Ubuntu 20.04 and I followed the kifarunix tutorial, but I am getting the mentioned error

**Comment by ltning** [comment]
>TLDR - when generating the certificates for the remote scanner, you must specify the CN so that it matches whatever IP/hostname/FQDN you will be connecting to the scanner from AND the certificate type on the remote scanner must be a client certificate, not a server certificate. If either of these are not configured correctly, you will receive SSL handshake errors as described above.

So you're saying the remote scanner - which is *receiving* connections from the `gvmd` process on the controlling node, should present a *client* certificate? How is that not in violation of key usage flags and basic TLS?

I really, really don't understand this - I'm doing TLS stuff for a living but gvmd <-> ospd-openvas completely escapes me.

**Comment by ltning** [comment]
> > TLDR - when generating the certificates for the remote scanner, you must specify the CN so that it matches whatever IP/hostname/FQDN you will be connecting to the scanner from AND the certificate type on the remote scanner must be a client certificate, not a server certificate. If either of these are not configured correctly, you will receive SSL handshake errors as described above.
> 
> So you're saying the remote scanner - which is _receiving_ connections from the `gvmd` process on the controlling node, should present a _client_ certificate? How is that not in violation of key usage flags and basic TLS?
> 
> I really, really don't understand this - I'm doing TLS stuff for a living but gvmd <-> ospd-openvas completely escapes me.

Replying to self - I can indeed confirm it works this way, but the log saying TLS key usage violation is patently wrong and misleading.

**Comment by ltning** [comment]
> > > TLDR - when generating the certificates for the remote scanner, you must specify the CN so that it matches whatever IP/hostname/FQDN you will be connecting to the scanner from AND the certificate type on the remote scanner must be a client certificate, not a server certificate. If either of these are not configured correctly, you will receive SSL handshake errors as described above.
> > 
> > 
> > So you're saying the remote scanner - which is _receiving_ connections from the `gvmd` process on the controlling node, should present a _client_ certificate? How is that not in violation of key usage flags and basic TLS?
> > I really, really don't understand this - I'm doing TLS stuff for a living but gvmd <-> ospd-openvas completely escapes me.
> 
> Replying to self - I can indeed confirm it works this way, but the log saying TLS key usage violation is patently wrong and misleading.

....aaaand it's still presenting the same (server) cert when connecting with `openssl s_client` (a server certificate). So what the `-k` and `-c` options actually do is uncertain...

**Comment by bjoernricks** [comment]
Using TLS certificates for the connection between ospd-openvas and gvmd is in an unmaintained state. We are not supporting this feature nor is it used in our products. Feel free to provide changes in pull requests to fix the unclear usage of TLS certs.
